### PR TITLE
Enable C++ deps pruning on Windows when PARSE_SHOWINCLUDES is available.

### LINKED ISSUE #14947: C++ dependency pruning is broken for Windows/MSVC
### Description of the problem:

With `gcc` style compiler `bazel` performs "dotD" dependency pruning of unused headers to limit recompilations if these unused headers were touched. As a feature this is somewhat documented in [`CODEBASE.md`](https://github.com/bazelbuild/bazel/blob/master/CODEBASE.md#input-discovery-and-input-pruning) and several issues here/mails on bazel-discuss.

It is less clear if `bazel` does/is supposed to do the same with MSVC toolchains. It is clear that `bazel` uses `/showIncludes` to perform [header inclusion checking](https://docs.bazel.build/versions/main/be/c-cpp.html#hdrs), but unclear if it goes further. https://github.com/bazelbuild/bazel/issues/11765 hints at header pruning on Windows being a feature.

### Bugs: what's the simplest, easiest way to reproduce this bug?

- Clone https://github.com/lummax/bazel-header-recompilation-reproducer
- run `bazel build :main -s`
- modify `dead.h` which may be used by `main` but is never included from `main.cpp`
- rerun `bazel build :main -s`

On Linux/gcc the second build does nothing, on Windows/MSVC the compilation unit is rebuild.

### What operating system are you running Bazel on?

Ubuntu 21.10 / Windows 10

### What's the output of `bazel info release`?

`release 5.0.0`

### Any other information, logs, or outputs that you want to share?

The following change in https://github.com/bazelbuild/bazel/blob/6c955a581fe880ddfde438d786390cebe00435bd/src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java#L350 fixes the problem, but I am not sure if that modification is safe.

```patch
 .../devtools/build/lib/rules/cpp/CppCompileActionBuilder.java   | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java b/src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java
index 36c81dc5..4472d771 100644
--- a/src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java
+++ b/src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java
@@ -347,7 +347,7 @@ public class CppCompileActionBuilder {
     if (grepIncludes != null) {
       realMandatoryInputsBuilder.add(grepIncludes);
     }
-    if (!shouldScanIncludes && dotdFile == null) {
+    if (!shouldScanIncludes && dotdFile == null && !featureConfiguration.isEnabled(CppRuleClasses.PARSE_SHOWINCLUDES)) {
       realMandatoryInputsBuilder.addTransitive(ccCompilationContext.getDeclaredIncludeSrcs());
       realMandatoryInputsBuilder.addTransitive(additionalPrunableHeaders);
     }
```

### PR DESCRIPTION
**Pull request by konste**
This fixes #14947.

### CODE REVIEW

**Comment by lberki** [comment]
I'll bounce this to @oquenchil and @meteorcloudy since they were more closely involved. Let me know if y'all need me to take a look at this.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java [hunk 1]
```diff
@@ -361,7 +365,7 @@ NestedSet<Artifact> buildMandatoryInputs() {
     if (grepIncludes != null) {
       realMandatoryInputsBuilder.add(grepIncludes);
     }
-    if (!shouldScanIncludes && dotdFile == null) {
+    if (!shouldScanIncludes && dotdFile == null && !shouldParseShowIncludes()) {
```
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java:368]: @oquenchil Do you know why `dotdFile` has to be null here?

Comment (konste) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileActionBuilder.java:368]: Those are conditions when graph trimming should not be used. With the compilers supporting dotd files it is when dotd file is not available. For MSVC it is when ParseShowIncludes is turned off. So this condition says "if there is no dotd and also no ParseShowIncludes then unfortunately do not trim the graph"

**Comment by konste** [comment]
@oquenchil this simple PR is a major blocker for us and badly needs your attention. Could you please merge it?

**Review by oquenchil — APPROVED** [review]

**Comment by kkpattern** [comment]
Hi, is it possible to provide a mechanism to disable pruning on Windows manually?

The main problem is that the `/showIncludes` output parsing is unreliable on Windows. There are two main problems:

1. MSVC outputs the absolute path of the header files. When remote execution is involved, it's tricky to convert the path back to the relative path.
2. Because of encoding and localization, it's hard for bazel to match the prefix of the `/showIncludes` output. For example, in our environment, the encoding is cp936, not utf-8. It's not feasible for bazel to try all the encodings plus all the languages MSVC may output.

If bazel fails to parse the `/showIncludes` output correctly, then when the header files are changed, the target won't get re-compiled, causing an incorrect build. This happened to us after we upgraded from 5.4.1 to 6.3.2

We also couldn't remove the `parse_showincludes` feature because then bazel will ask for the dotD file, which MSVC won't produce.

I have considered the following options：

1. update `isGenerateDotdFile` method, adding an additional feature to tell bazel that the current toolchain won't produce the dotD file. Then we can disable pruning by removing the parse_showincludes feature.
2. Add a command line flag to disable pruning.

If one of the above solutions is feasible, or there is a better solution, we are willing to help prepare a PR.

Thanks.

**Comment by meteorcloudy** [comment]
I believe we can make option 1 work, /cc @buildbreaker2021 as the C++ rule owner.

**Comment by kkpattern** [comment]
I did some experiments:

1. We can add `--action_env=VSLANG=1033` to force MSVC to generate output in English. This way, Bazel can parse the paths correctly, no matter their encoding.
2. For remote execution, if we find `/showIncludes` in the action command parameters, we convert the output paths to paths relative to the working directory in the remote runner. Then, pruning works perfectly.

Though we can workaround the issue without changing Bazel itself, I think it's important to prevent pruning from causing wrong builds without the user's knowledge.

Since Bazel can only understand UTF-8 encoding, I think we can try decoding the output with UTF-8. If it fails, warn the user that pruning will not work correctly. Then, the user can choose to either change the encoding or turn off the parse_showincludes feature.

Should I create a post on bazel-discuss and continue the discussion there?

**Comment by meteorcloudy** [comment]
> We can add `--action_env=VSLANG=1033` to force MSVC to generate output in English.

Nice finding! I guess we can add this env var as a default to the `parse_showincludes` feature? 
https://github.com/bazelbuild/bazel/blob/b27ca73617c5319d4b15a9e5ba499cf476679a95/tools/cpp/windows_cc_toolchain_config.bzl#L791

For remote execution, I remember we have some trick to ignore the execroot prefix, not sure why that doesn't work.
https://github.com/bazelbuild/bazel/blob/b27ca73617c5319d4b15a9e5ba499cf476679a95/src/main/java/com/google/devtools/build/lib/rules/cpp/ShowIncludesFilter.java#L167-L180

> Should I create a post on bazel-discuss and continue the discussion there?

Sure, we recommend to use https://github.com/bazelbuild/bazel/discussions now.

**Comment by kkpattern** [comment]
I have created a discussion here: https://github.com/bazelbuild/bazel/discussions/19451.

**Pull request closed without merging**